src/main.cpp

```diff
@@ -6398,6 +6098,26 @@ bool ProcessMessage(CNode *pfrom, std::string strCommand, CDataStream &vRecv, in
             LOG(NET, "more getheaders (%d) to end to peer=%s (startheight:%d)\n", pindexLast->nHeight,
                 pfrom->GetLogName(), pfrom->nStartingHeight);
             pfrom->PushMessage(NetMsgType::GETHEADERS, chainActive.GetLocator(pindexLast), uint256());
+
+            // During the process of IBD we need to update block availability for every connected peer. To do that we
+            // request, from each NODE_NETWORK peer, a header that matches the last blockhash found in this recent set
+            // of headers. Once the reqeusted header is received then the block availability for this peer will get
+            // updated.
+            if (IsInitialBlockDownload())
+            {
+                LOCK(cs_vNodes);
+                for (CNode *pnode : vNodes)
+                {
+                    if (!pnode->fClient && pnode != pfrom)
+                    {
+                        // We only want one single header so we pass a null for CBlockLocator.
```

Comment:
If the block has already given us its best headers, this PushMessage might be unnecessary.  I am also worried about an infinite request loop.  right now this can't happen because PushMessage only asks for a single header, but this code only triggers if nCount == MAX_HEADERS_RESULTS.  But the fact that this prevents an request loop is a secondary effect that might be missed by a future developer.  We can solve this by checking that the node doesn't already have a better header here.  I see that check in UpdateBlockAvailability:

if (state->pindexBestKnownBlock == NULL || it->second->nChainWork >= state->pindexBestKnownBlock->nChainWork)
            state->pindexBestKnownBlock = it->second;

Does it make sense to do a similar check here and not send a GETHEADERS if our best block already has greater work than this one?

Reply:
hmmm, yes i think you're right, but the problem is cs_main again and locking order...will have to create a copy of vnodes, update refs, release ref's etc...it really bugs me that state is dependent on cs_main...shouldn't have to be...i'll look at that at another time, but for now i'll work on your requested change.

Reply:
@gandrewstone ok, done...